keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,720 @@
+masquerade or drop a packet, it just marks it as needing to be
+masqueraded or dropped later, and then one of the later chains handles
+the actual masquerading/dropping.
+
+This approach was not _necessary_; kube-proxy could have just
+duplicated its matching logic between multiple IPTables chains, eg,
+having a rule saying "a packet sent to `192.168.0.11:80` should be
+redirected to `10.0.20.18:8080`" in one chain and a rule saying "a
+packet whose original pre-NAT destination was `192.168.0.11:80` should
+be masqueraded" in another.
+
+But using `KUBE-MARK-MASQ` allows kube-proxy to keep its logic more
+efficient, compact, and well-organized than it would be otherwise; it
+can just create a pair of adjacent rules saying "a packet sent to
+`192.168.0.11:80` should be redirected to `10.0.20.18:8080`, and
+should also be masqueraded".
+
+#### Eliminating `KUBE-MARK-DROP`
+
+In theory, `KUBE-MARK-DROP` is useful for the same reason as
+`KUBE-MARK-MASQ`, although in practice, it turns out to be
+unnecessary. Kube-proxy currently drops packets in two cases:
+
+  - When using the `LoadBalancerSourceRanges` feature on a service,
+    any packet to that service that comes from outside the valid
+    source IP ranges is dropped.
+
+  - When using `Local` external traffic policy, if a connection to a
+    service arrives on a node that has no local endpoints for that
+    service, it is dropped.
+
+In both cases, the use of `KUBE-MARK-DROP` from the `nat` table could
+easily be replaced with direct calls to `-j DROP` in the `filter`
+table.
+
+In the `LoadBalancerSourceRanges` case, we currently create a
+per-service `KUBE-FW` chain in the `nat` table that checks against
+each allowed source range and forwards traffic from those ranges to
+the `KUBE-SVC` or `KUBE-XLB` chain for the service. At the end
+of the chain, it calls `KUBE-MARK-DROP` on any unmatched packets.
+
+But there is no reason why the `FW` chain could not simply be moved to
+the `filter` table (with calls to `-j ACCEPT` for allowed source
+ranges and `-j DROP` for everything else). Then in the `nat` table,
+the rule matching the LoadBalancer IP could forward traffic directly
+to the `SVC` or `XLB` chain, just like the NodePort and ExternalIP
+matches do, because the disallowed ranges will already have been
+filtered out.
+
+(There are comments in the kube-proxy code suggesting that at one
+point there was a plan to make NodePort connections obey
+`LoadBalancerSourceRanges` as well. This may be why the firewalling
+was originally implemented in the `nat` table; to avoid duplicating
+both the LoadBalancer-matching logic and the NodePort-matching logic
+in `filter` as well. However, this was never implemented, and no one
+ever seemed to care. Given the current state, in which only
+LoadBalancer traffic gets firewalled, the simplification from reusing
+the existing matching logic in the `filter` table seems to be
+outweighed by the complication of having to deal with
```

Comment:
I'm 50/50 - having MARK-DROP in the nat chain puts all the logic in one place, which is nice.  That said, I buy the REJECT consistency below